### ISSUE TITLE
New rule: Prefer const array "as const"

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#3259]
If an array isn't changed nor exported, it should be defined `as const` where possible. I haven't thought this through yet, but it feels similar to `no-let`. The same applies to objects 

## Fail

```ts
function getIndex(index): number | undefined {
	const list = [1, 2, 3]; // This is a constant array, why isn't it `as const`?
	return list[index];
}
```
```ts
const list = [1, 2, 3]; // This is never modified, shouldn't it be `as const`?
function getIndex(index): number | undefined {
	return list[index];
}
```

## Pass

```ts
function getIndex(index): number | undefined {
	const list = [1, 2, 3] as const; // as const, perfect
	return list[index];
}
```
```ts
function getIndex(index): number | undefined {
	let list = [1, 2, 3]; // let
	return list[index];
}
```
```ts
const list = [1, 2, 3] as const;
function getList(index): number | undefined {
	return list; // Can't guarantee that it won't be changed outside
}
```

```ts
const list = [1, 2, 3];
function getIndex(index): number | undefined {
	list.splice(0, 1, index); // Array is being modified
	return list[index];
}
```

**Comment by fregante** [comment]
And of course I'm regretting this as it brings some unwanted breakage, but  perhaps that can be fixed otherwise.
<img width="1186" alt="Screen Shot 2021-04-02 at 14 59 55" src="https://user-images.githubusercontent.com/1402241/113395482-1fb8a600-93c4-11eb-8e51-4bf204a276f5.png">

**Comment by bradzacher** [comment]
Whilst I agree with this in spirit - readonly everything is a much better default.
In practice I disagree with it because it makes so many things hard.

As you mentioned - the types are just all wrong and it makes it really difficult to do many things with the tuple.

Using `as const` arrays also causes you issues if you pass it into a `Set`:
```ts
const a = [1,2,3];
const b = new Set(a); // typeof b === Set<number>

const x = [1,2,3] as const;
const y = new Set(x); // typeof y === Set<1 | 2 | 3>
```

This is also just a really hard to build lint rule because TS does not track *values*, only types.

For example:
```ts
const x = [1,2,3];

const y = x;

y.push(4);

// has `x` been mutated????
```

There's also the issue of what you do across function boundaries:

```ts
function get(arr: readonly number[]: idx: number): number {
  return arr[idx];
}

const x = [1,2,3];
get(x, 1);

// has `x` been mutated????
```
There's no way to track this in TS because there's no mechanism to understand what a function does - TS lacks any way to declare the purity of a function. This would lead to false-negatives.

you *could* solve this by using type information to understand that `get -> arr` was declared as `readonly`, but that assumes library authors are doing the *right* thing (which most arent)

**Comment by fregante** [comment]
> Using `as const` arrays also causes you issues if you pass it into a `Set`:

Indeed, this kind of issues kills it already for me. `as const` remains as a nice feature — when it can be used.
